### PR TITLE
added password sign in function for users

### PR DESCRIPTION
**Pull request by injustli**

### CODE REVIEW

**Review by chilimalone — COMMENTED** [review]
Mostly looks ok. Added some interesting reads for you. I think we should at least hash the passwords. We can leave the salting to another pull request, after you've had some time to soak how security and passwords work.

**Comment by Lord-Bubbles** [comment]
currently with the way i have it, once the user adds himself, he must sign in. If you add other users without signing in, only the current user can sign in and old one can no longer do so. Probably should change that right?

File: src/codeu/chat/client/ClientUser.java [hunk 1]
```diff
@@ -84,6 +87,15 @@ public boolean signOutUser() {
     return hadCurrent;
   }
 
+  public boolean checkPassword(String password) {
+    String test = password + salt;
+    int testhash = test.hashCode();
```
Comment (chilimalone) [inline src/codeu/chat/client/ClientUser.java]: let's use SHA-256 - https://www.baeldung.com/sha-256-hashing-java
So.... MessageDigest sha-256 hash

File: src/codeu/chat/client/ClientUser.java [hunk 2]
```diff
@@ -119,6 +131,27 @@ public void addNickname(String name) {
...
+        "abcdefghijklmnopqrstuvwxyz";
+    StringBuilder sb = new StringBuilder(6);
+    for (int i = 0; i < 6; i++) {
+      int index = (int) (alphaNumeric.length() * Math.random());
+      sb.append(alphaNumeric.charAt(index));
+    }
+    salt = sb.toString();
+    String password = phrase + salt;
+    hashstring.put(salt, password.hashCode());
+
```
Comment (chilimalone) [inline src/codeu/chat/client/ClientUser.java]: nit: remove extra blank line

**Review by chilimalone — COMMENTED** [review]
Where is the server-side change?

i.e. serializing the user and actually storing it into the database?

File: src/codeu/chat/client/ClientUser.java [hunk 3]
```diff
@@ -14,10 +14,12 @@
 
 package codeu.chat.client;
 
-import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Arrays;
```
Comment (chilimalone) [inline src/codeu/chat/client/ClientUser.java]: alphabetize

File: src/codeu/chat/client/ClientUser.java [hunk 4]
```diff
@@ -14,10 +14,12 @@
...
+import java.security.*;
```
Comment (chilimalone) [inline src/codeu/chat/client/ClientUser.java]: try not to import * unless you need more than 100 files. This is for code longevity - if oracle decides to add a class called User in java.security, then all of a sudden your code will stop compiling...

File: src/codeu/chat/client/ClientUser.java [hunk 5]
```diff
@@ -14,10 +14,12 @@
...
+import java.nio.charset.*;
```
Comment (chilimalone) [inline src/codeu/chat/client/ClientUser.java]: ditto

File: src/codeu/chat/client/ClientUser.java [hunk 6]
```diff
@@ -34,6 +36,8 @@
 
   private User current = null;
 
+  private int counter = 0;
```
Comment (chilimalone) [inline src/codeu/chat/client/ClientUser.java]: I don't see counter being used?

File: src/codeu/chat/client/ClientUser.java [hunk 7]
```diff
@@ -84,6 +89,19 @@ public boolean signOutUser() {
     return hadCurrent;
   }
 
+  public void setCurrent(User user) {
+    current = user;
+  }
+
+  public boolean checkPassword(String password) {
+    String test = password + getCurrent().getSalt();
```
Comment (chilimalone) [inline src/codeu/chat/client/ClientUser.java]: so this is why i don't think you should verify this on the client side - wouldn't getCurrent() be null when you first open the chat client? How do you know which user is going to log in without figuring out the username?

**Review by chilimalone — COMMENTED** [review]
So close!

File: src/codeu/chat/client/ClientUser.java [hunk 8]
```diff
@@ -89,14 +86,10 @@ public boolean signOutUser() {
...
-    current = user;
-  }
-
-  public boolean checkPassword(String password) {
-    String test = password + getCurrent().getSalt();
+  public boolean checkPassword(String name, String password) {
+    String test = password + usersByName.first(name).getSalt();
     String hexstring = hexSHA(test);
-    if (!hexstring.equals(getCurrent().getHash())) {
+    if (!hexstring.equals(usersByName.first(name).getHash())) {
```
Comment (chilimalone) [inline src/codeu/chat/client/ClientUser.java]: when creating a user, should we initialize a hash?

File: src/codeu/chat/client/simplegui/UserPanel.java [hunk 9]
```diff
@@ -158,7 +158,7 @@ public void actionPerformed(ActionEvent e) {
             final String s = (String) JOptionPane.showInputDialog(
                 UserPanel.this, "Enter password: ", "Enter Password", JOptionPane.PLAIN_MESSAGE, 
                 null, null, "");
-            if (s != null && clientContext.user.checkPassword(s)) {
+            if (s != null &&  clientContext.user.checkPassword(data,s)) {
```
Comment (chilimalone) [inline src/codeu/chat/client/simplegui/UserPanel.java]: nit: extra space between && and clientContext; add space between data, and 's'

File: src/codeu/chat/client/simplegui/UserPanel.java [hunk 10]
```diff
@@ -182,21 +182,21 @@ public void actionPerformed(ActionEvent e) {
...
+            final String x = (String) JOptionPane.showInputDialog(
+                UserPanel.this, "Enter password:", "Add Password", JOptionPane.PLAIN_MESSAGE,
+                null, null, "");
+            if (clientContext.user.isValidPassword(x)) {
+              clientContext.user.addPassword(s,x);
+              break;
+            } else if (x == null) {
+              break;
+            } else {
+              JOptionPane.showMessageDialog(UserPanel.this, "Password isn't at least 6 characters long!",
```
Comment (chilimalone) [inline src/codeu/chat/client/simplegui/UserPanel.java]: How about "Passwords should be at least 6 characters long"

**Review by chilimalone — APPROVED** [review]
A few small nits - please fix before submitting. Probably your next change should be updating the user creation stuff so that users can be created with passwords

**Pull request merged**